Framework/API/src/ExperimentInfo.cpp

```diff
@@ -300,12 +306,17 @@ void ExperimentInfo::setInstrument(const Instrument_const_sptr &instr) {
   m_detectorInfoWrapper = Kernel::make_unique<DetectorInfo>(
       *m_detectorInfo, getInstrument(), m_parmap.get());
 
-  std::tie(m_componentInfo, m_componentIds) =
+  std::tie(m_componentInfo, m_componentIds, m_componentIdToIndexMap) =
       makeComponentInfo(*instr, detectorInfo());
   m_parmap->setComponentInfo(m_componentInfo);
-  m_componentInfoWrapper =
-      Kernel::make_unique<ComponentInfo>(*m_componentInfo, m_componentIds);
-  // Detector IDs that were previously dropped because they were not part of the
+
+    /*
+     * If the ID -> index map has already been created. Reuse it.
+     * */
+    m_componentInfoWrapper = Kernel::make_unique<ComponentInfo>(
+        *m_componentInfo, m_componentIds, m_componentIdToIndexMap);
```

Comment:
I am not quite sure I understand why all these changes to `ExperimentInfo` are necessary. Can't you keep creating the IDs in `API::ComponentInfo` and simple extract them in `getInstrument()`, and here cover 2 cases:
- If the instrument has `ComponentInfo` (and thus the map), use the new constructor you added to `API::ComponentInfo`.
- Otherwise, use the older constructor and let `ComponentInfo` build the map.